File: ZKCache/src/main/java/com/ericsson/ema/tim/dml/predicate/Like.java
package com.ericsson.ema.tim.dml.predicate;

import com.ericsson.ema.tim.dml.DataTypes;
import com.ericsson.ema.tim.dml.DmlBadSyntaxException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

public class Like extends AbstractPredicate implements Predicate {
    private final static Logger LOGGER = LoggerFactory.getLogger(Eq.class);

    private Like(String field, String value) {
        super(field, value);
    }

    public static Like like(String field, String value) {
        return new Like(field, value);
    }

    @Override
    public boolean eval(Object tuple) {
        if (this.valueToComp == null)
            return false;

        Object fieldVal = getFiledValFromTupleByName(tuple);
        Map<String, String> metadata = getSelector().getContext().getTableMetadata();
        String fieldType = metadata.get(field);
        switch (fieldType) {
            case DataTypes.String:
                return ((String) fieldVal).contains((String) this.valueToComp);
            default:
                LOGGER.error("unsupported data type: {},{}", field, fieldType);
                throw new DmlBadSyntaxException("unsupported data type: " + field + "," + fieldType);
        }
    }
}

File: zkcache_scala/src/main/scala/com/ericsson/ema/tim/dml/predicate/Like.scala
package com.ericsson.ema.tim.dml.predicate

import com.ericsson.ema.tim.dml.DataTypes
import com.ericsson.ema.tim.exception.{DmlBadSyntaxException, DmlNoSuchFieldException}
import org.slf4j.LoggerFactory

/**
  * Created by eqinson on 2017/5/12.
  */
class Like private(override val field: String, override val valueToComp: Object) extends AbstractPredicate(field, valueToComp) with Predicate {
	private val LOGGER = LoggerFactory.getLogger(classOf[Like])

	override def eval(tuple: Object): Boolean = {
		if (this.valueToComp eq null)
			return false

		val fieldVal: Object = getFiledValFromTupleByName(tuple)
		val fieldType = selector.context.tableMetadata.get(field)
		fieldType match {
			case Some(DataTypes.String) =>
				fieldVal.asInstanceOf[String].matches(valueToComp.asInstanceOf[String])
			case Some(_)                =>
				LOGGER.error("unsupported data type: {},{}", field, fieldType: Any)
				throw DmlBadSyntaxException("unsupported data type: " + field + "," + fieldType)
			case None                   => throw DmlNoSuchFieldException(field)
		}
	}
}

object Like {
	def apply(field: String, value: String) = new Like(field, value)
}
